R/as-data-frame.R

```diff
@@ -44,38 +44,62 @@ as.data.frame.CrunchDataset <- function (x, row.names = NULL, optional = FALSE,
 }
 
 #' @rdname dataset-to-R
+#' @importFrom utils read.csv
 #' @export
 as.data.frame.CrunchDataFrame <- function (x, row.names = NULL, optional = FALSE, ...) {
     ds <- attr(x, "crunchDataset")
-    default.stringsAsFactors <- function () FALSE
-    limit <- min(c(10000, getOption("crunch.data.frame.limit")))
-    if (nrow(ds) * ncol(ds) > limit) {
-        ## TODO: switch to downloading CSV and reading that?
-        halt("Dataset too large to coerce to data.frame. ",
-            "Consider subsetting it first")
-    }
-    var_names <- names(x)
-    # TODO: something intelligent with modes
-    out <- lapply(var_names, function(var) {
-        values <- x[[var]]
-        
-        # Flatten all array variables, using only the subvariable name. We need 
-        # to create a list layer for non-dataframe values so that we can unlist 
-        # them later.
-        if (is.data.frame(values)) {
-            values <- as.list(values)
+    tmp <- tempfile()
+    write.csv(ds, tmp)
+    ds_out <- read.csv(tmp, stringsAsFactors = FALSE)
+    ds_out[] <- unlist(lapply(ds, coerceVariable, ds_out), recursive = FALSE)
+
+    var_names <- lapply(names(x), function(v){
+        if (is.Array(ds[[v]])) {
+            return(names(subvariables(ds[[v]])))
+        } else {
+            return(v)
+        }
+    })
+    var_names <- unlist(var_names)
+
+    ## Crunch Dataframes contain both server variables and local variables this
+    ## ensures that both are returned in the proper order.
+    out <- lapply(var_names, function(v){
+        if (v %in% names(ds_out)) {
+            if (v %in% names(ds) && tuple(ds[[v]])$discarded) {
+                warning("Variable ", v, " is hidden", call.=FALSE)
+            }
+            return(ds_out[[v]])
         } else {
-            values <- list(values)
-            names(values) <- var
+            return(x[[v]])
         }
-        
-        return(values)
     })
-    out <- unlist(out, recursive = FALSE)
-    
+    names(out) <- var_names
     return(structure(out, class="data.frame", row.names=c(NA, -nrow(ds))))
 }
 
+coerceVariable <- function(var, df){
+    coerceFactor <- function(x){
+        na_cats <- is.na(categories(var))
+        v <- as.vector(x)
+        v[v %in% na_cats] <- NA
+        v <- factor(v, levels = names(categories(var))[!na_cats])
+        return(v)
```

Comment:
It doesn't look like this would work if someone changed the mode that categoricals are being pulled down as. What if someone set `categorical.mode` on the CrunchDataFrame to be numeric or id?

Also, I would put this function definition outside the `coerceVariable` function, so that you can unit test behaviors like that more easily.